docs: add halo2 page (#15)
anonklub/anonklub.github.io: docs/prove/halo2.md

> Refer to the [Halo2 circuit in `anonklub/anonklub` repository](https://github.com/anonklub/anonklub/tree/main/pkgs)

> Requirements
>
> - [ ] Install [@anonklub/halo2-binary-merkle-tree-worker](#), a web worker that includes the WebAssembly compilation of the Halo2 Merkle tree gadget.
> - [ ] Install [@anonklub/halo2-eth-membership-worker](#), a web worker that contains the WebAssembly compilation of the Halo2 circuit.

## TLDR

### 

```js
// React.js / Next.js example
import {
    Halo2EthMembershipWorker,
    type ProveMembershipFn,
    type VerifyMembershipFn,
} from '@anonklub/halo2-eth-membership-worker'
import {
  type GenerateMerkleProofFn,
  Halo2BinaryMerkleTreeWorker,
} from '@anonklub/halo2-binary-merkle-tree-worker'
import { useEffect, useState } from 'react'

export const useWorker = (
  worker:typeof Halo2EthMembershipWorker,
) => {
  const [isWorkerReady, setIsWorkerReady] = useState(false)

  useEffect(() => {
    void (async () => {
      await worker.prepare()
      setIsWorkerReady(true)
    })()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  return isWorkerReady

> }

Reviewer: can be removed, covered in page added by #17 now right?

export const useHalo2BinaryMerkleTreeWorker = () => {
  const isWorkerReady = useWorker(Halo2BinaryMerkleTreeWorker)

  const generateMerkleProof: GenerateMerkleProofFn = async (
    leaves,
    leaf,
    depth,
  ): Promise<Uint8Array> => {
    process.env.NODE_ENV === 'development' && console.time('==>merkle')

    // eslint-disable-next-line no-useless-catch

>     try {

Reviewer: does not reflect our actual code (no try/catch)

  }

  return {
    isWorkerReady,
    proveMembership,
    verifyMembership,
  }
}
```

## Step By Step

### Prepare

[@anonklub/halo2-binary-merkle-tree-worker](#) and [@anonklub/halo2-eth-membership-worker](#) are designed to operate on the client side. In the example above, ensure that you run prepare from each worker using `await worker.prepare()`. This function initializes the WebAssembly (WASM) circuit and determines the number of available threads in the browser to initialize the thread pool:

#### `useHalo2BinaryMerkleTreeWorker.prepare()`

```js
async prepare() {
    halo2BinaryMerkleTreeWasm = await import(
        '@anonklub/halo2-binary-merkle-tree/dist/'
    )

    const wasmModuleUrl = new URL(
        '@anonklub/halo2-binary-merkle-tree/dist/index_bg.wasm',
        import.meta.url,
    )
    const response = await fetch(wasmModuleUrl)
    const bufferSource = await response.arrayBuffer()

    await halo2BinaryMerkleTreeWasm.initSync(bufferSource)
    await halo2BinaryMerkleTreeWasm.initPanicHook()
    const numThreads = navigator.hardwareConcurrency
    await halo2BinaryMerkleTreeWasm.initThreadPool(numThreads)
}
```

#### `useHalo2EthMembershipWorker.prepare()`

> ```js

Reviewer: can be removed too no?
If needed refer to the wasm page

> Refer to the [Halo2 circuit in `anonklub/anonklub` repository](https://github.com/anonklub/anonklub/tree/main/pkgs/halo2-eth-membership)

> Requirements
>
> - [ ] Install [@anonklub/halo2-binary-merkle-tree-worker](https://www.npmjs.com/package/@anonklub/halo2-binary-merkle-tree-worker), a web worker that includes the WebAssembly compilation of a [binary merkle tree rust implmentation with a Halo2 gadget for merkle proof verification](https://github.com/anonklub/anonklub/tree/main/pkgs/halo2-binary-merkle-tree).
> - [ ] Install [@anonklub/halo2-eth-membership-worker](https://www.npmjs.com/package/@anonklub/halo2-eth-membership-worker), a web worker that contains the WebAssembly compilation of the [Halo2 circuit](https://github.com/anonklub/anonklub/tree/main/pkgs/halo2-eth-membership).

## TLDR

> 
> ### 

Reviewer: ```suggestion
```
This messes up the style of the side bar

![image](https://github.com/user-attachments/assets/efea4c19-11e2-4264-b56f-8fdc8e89d119)

Reviewer: Same for the spartan page.
Can you fix it there too?

4. The generated proof will be serialized, making it immediately usable as a parameter for the proof function in the circuit.

```js
export type GenerateMerkleProofFn = (
  leaves: string[],
  leaf: string,
  depth: number,
) => Promise<Uint8Array>
```

## Proof of Membership

Once you have generated the Merkle proof for an Ethereum address (leaf), you can proceed to create a Halo2 proof for the membership of that address in the Merkle tree. Follow these steps to generate a Halo2 membership proof:

1. Call the `prepare()` function as outlined earlier.
2. Sign a `message`and obtain the `signature` in hexadecimal format.
3. The generated Halo2 membership proof `membershipProofSerialized` is in a serialized form and it will be ready for immediate use in the `verifyMembership()` function.

```js
export interface ProveInputs {
  sig: Hex
  message: string
  merkleProofBytesSerialized: Uint8Array,
}
```

### Example of use

```js
import { useAsync } from 'react-use'
import type { Hex } from 'viem'

> import { useHalo2EthMembershipWorker } from '@/hooks/useHalo2EthMembershipWorker'

Reviewer: what is `@/hook`?
I know what this is, but other devs won't.  Because it is a custom alias used by our UI.

> import { useStore } from '@/hooks/useStore'

Reviewer: what is `useStore`?
I know what it is, but other devs won't, because this is a custom hook specific to our UI implementation.
Either remove it or link to something that explains what it is.

export const useProofResult = () => {

>   const { proofRequest } = useStore()

Reviewer: what is `proofRequest`?
Again I know, but it is not described in this page.
Maybe here we can link to https://anonklub.github.io/#/prove/circom?id=create-proof-request

  const { isWorkerReady, proveMembership } = useHalo2EthMembershipWorker()

  return useAsync(async () => {
    if (proofRequest === null || !isWorkerReady) return

    return await proveMembership({
      merkleProofBytesSerialized: proofRequest.merkleProof,
      message: proofRequest.message,
      sig: proofRequest.rawSignature as Hex,
    })
  }, [isWorkerReady, proofRequest])
}
```

## Membership Verification

After successfully generating the Halo2 proof, you can proceed with verifying that proof in Halo2. Follow these steps:

- Ensure you have the `membershipProofSerialized` output from the proof of membership step.

```js
export interface VerifyInputs {
anonklubProof: Uint8Array
}
```

### Example of use

```js
import { useHalo2EthMembershipWorker } from '@/hooks/useHalo2EthMembershipWorker';

> import { useStore } from '@/hooks/useStore';

Reviewer: same here,
sudden use of custom alias and custom hooks that may confuse devs because they are not documented in this page or anywhere in the docs website.

    const proof = await Halo2BinaryMerkleTreeWorker.generateMerkleProof(
        leaves,
        leaf,
        depth,
    )

    process.env.NODE_ENV === 'development' && console.timeEnd('==>merkle')
    return proof
  }

  return {
    generateMerkleProof,
    isWorkerReady,
  }
}

// A React hook example for integrating with `@anonklub/halo2-eth-membership-worker` circuit web-worker; https://github.com/anonklub/anonklub/blob/main/ui/src/hooks/useHalo2EthMembershipWorker.ts
export const useHalo2EthMembershipWorker = () => {
  const isWorkerReady = useWorker(Halo2EthMembershipWorker)

  const proveMembership: ProveMembershipFn = async ({
    sig,
    message,
    merkleProofBytesSerialized,
    k
  }): Promise<Uint8Array> => {
    process.env.NODE_ENV === 'development' && console.time('==> Prove')

    const proof = await Halo2EthMembershipWorker.proveMembership({
      merkleProofBytesSerialized,
      message,
      sig,
    })

    process.env.NODE_ENV === 'development' && console.timeEnd('==> Prove')

    return proof
  }

  const verifyMembership: VerifyMembershipFn = async ({
    membershipProofSerialized,
    k

>   }): Promise<boolean> => {

Reviewer: i think we should provide more information about what `k` is.
If I was a reader new to the project, I wouldn't know what `k` is and what value to choose.
Maybe you should link to the benchmark section here? And make clear that based on the results of your benchmark, you recommend using `k=14` (what the UI does)

Author: That is a typo actually, we dont use `k` anymore as params to the worker since it is fixed in the worker page.I added in the Benchmark that we are using KZG on degree of Polynomial of 14 `K`, with a description for both of them.

    process.env.NODE_ENV === 'development' && console.time('==> Verify')

    const isVerified = await Halo2EthMembershipWorker.verifyMembership({
      membershipProofSerialized,
      k
    })

    process.env.NODE_ENV === 'development' && console.timeEnd('==> Verify')

    return isVerified
  }

  return {
    isWorkerReady,
    proveMembership,
    verifyMembership,
  }
}
```

## Step By Step

### Prepare

> [@anonklub/halo2-binary-merkle-tree-worker](https://www.npmjs.com/package/@anonklub/halo2-binary-merkle-tree-worker) and [@anonklub/halo2-eth-membership-worker](https://www.npmjs.com/package/@anonklub/halo2-eth-membership-worker) are designed to operate on the client side. In the example above, ensure that you run prepare from each worker using `await worker.prepare()`. Please check [`Wasm & Web-Workers`](https://anonklub.github.io/#/prove/wasm) doc for more details.

Reviewer: ```suggestion
[@anonklub/halo2-binary-merkle-tree-worker](https://www.npmjs.com/package/@anonklub/halo2-binary-merkle-tree-worker) and [@anonklub/halo2-eth-membership-worker](https://www.npmjs.com/package/@anonklub/halo2-eth-membership-worker) are designed to operate on the client side. In the example above, ensure that prepare each worker using `await worker.prepare()`. Please check [`Wasm & Web-Workers`](https://anonklub.github.io/#/prove/wasm) doc for more details.
```

## Merkle Proof

Generating a Merkle proof to verify the inclusion of an Ethereum address within a Merkle tree is crucial for the circuit's functionality. We use a binary Merkle tree structure, and the `@anonklub/halo2-binary-merkle-tree-worker` library provides a [gadget](https://zcash.github.io/halo2/concepts/gadgets.html) that serves as a constraint for verifying the Merkle proof within the circuit. Follow these steps to generate a Merkle proof:

1. Call the `prepare()` function as previously described.

> 2. Prepare the list of Ethereum addresses. The Anonklub project can assist in scanning the blockchain to create this list, check out [query docs](https://anonklub.github.io/#/apis?id=query).

Reviewer: that is inaccurate or even incorrect.  
The query API does not scan the blockchain itself. It requests and post processes data from services that serve already scanned/indexed blockchain data.
```suggestion
2. Prepare the list of Ethereum addresses. The Anonklub project can assist in building this list by querying services that serve indexed blockchain data, check out [query docs](https://anonklub.github.io/#/apis?id=query).
```

Author: That is more accurate definition, I updated that in the Spartan page as well.

export const useProofResult = () => {
  const { proofRequest } = useStore()
  const { isWorkerReady, proveMembership } = useHalo2EthMembershipWorker()

  return useAsync(async () => {
    if (proofRequest === null || !isWorkerReady) return

    return await proveMembership({
      merkleProofBytesSerialized: proofRequest.merkleProof,
      message: proofRequest.message,
      sig: proofRequest.rawSignature as Hex,
    })
  }, [isWorkerReady, proofRequest])
}
```

## Membership Verification

After successfully generating the Halo2 proof, you can proceed with verifying that proof in Halo2. Follow these steps:

- Ensure you have the `membershipProofSerialized` output from the proof of membership step.

```js
export interface VerifyInputs {
anonklubProof: Uint8Array
}
```

### Example of use

```js
import { useHalo2EthMembershipWorker } from '@/hooks/useHalo2EthMembershipWorker';
import { useStore } from '@/hooks/useStore';
import { useAsync } from 'react-use';

export const useVerifyProof = () => {
  const { proof } = useStore();
  const { verifyMembership } = useHalo2EthMembershipWorker();

  return useAsync(async () => {
    if (proof === null) return;

>     return await verifyMembership(proof);

Reviewer: this signature is inconsistent with what is described in the TLDR section, 
```ts
const verifyMembership: VerifyMembershipFn = async ({
    membershipProofSerialized,
    k
  }): Promise<boolean>
```
this will confuse readers.
Stick to one.
I suggest in the docs we don't "hide" `k`. So we use everywhere `{ membershipProofSerialized, k }` as params

Reviewer: this is still not matching.
in the TLDR section you are using
```ts
const verifyMembership: VerifyMembershipFn = async ({ membershipProofSerialized }): Promise<boolean>
```

  }, [proof]);
};
```

> ## Benchmarks

Reviewer: :+1: 
Great to have thought about adding this section.

  // Source code: https://github.com/anonklub/anonklub/blob/main/ui/src/hooks/useProofRequest.ts
  // 
  // For more detailed information on proof requests,
  // refer to the @anonklub/proof documentation: https://anonklub.github.io/#/prove/circom?id=create-proof-request
  // refer to source code: https://github.com/anonklub/anonklub/blob/6884d934f95f2c153bd9531aca36ba7a6e6d4720/pkgs/proof/src/ProofRequest.ts
  const { proofRequest } = useStore()
  const { isWorkerReady, proveMembership } = useHalo2EthMembershipWorker()

  return useAsync(async () => {
    if (proofRequest === null || !isWorkerReady) return

    return await proveMembership({
      merkleProofBytesSerialized: proofRequest.merkleProof,
      message: proofRequest.message,
      sig: proofRequest.rawSignature as Hex,
    })
  }, [isWorkerReady, proofRequest])
}
```

## Membership Verification

After successfully generating the Halo2 proof, you can proceed with verifying that proof in Halo2. Follow these steps:

- Ensure you have the `membershipProofSerialized` output from the proof of membership step.

```js
export interface VerifyInputs {
anonklubProof: Uint8Array
}

> ```

Reviewer: this is still wrong.
inputs is just `somePositionalArg: Uint8Array` not `{ anonklubProof: Uint8Array }`
so you can't use an interface. You must use a `type` alias. Which is what we do is the UI.

https://github.com/anonklub/anonklub/blob/6884d934f95f2c153bd9531aca36ba7a6e6d4720/pkgs/halo2-eth-membership-worker/src/interface.ts#L22

### Example of use

```js
import { useHalo2EthMembershipWorker } from '@/hooks/useHalo2EthMembershipWorker';
import { useAsync } from 'react-use';
// useStore is a custom React hook for managing global state in the application
// using the `easy-peasy` state management library.`
//
// Hook Path: `@hook` alias points to `src/hooks` for convenient imports.//

> //

Reviewer: the import is `@/hooks`
but your comment is about `@hook`
It doesn't match.

In the UI we use neither: we use `@hooks`

Author: > In the UI we use neither: we use @hooks

In the UI we use `@/hooks` imports: 
- https://github.com/anonklub/anonklub/blob/6884d934f95f2c153bd9531aca36ba7a6e6d4720/ui/src/hooks/useHalo2EthMembershipWorker.ts#L1

So the typo is missing `/` in the comment

// `easy-peasy`: https://github.com/ctrlplusb/easy-peasy
//
// For more detailed info you can refer to the full source code here::
// https://github.com/anonklub/anonklub/blob/main/ui/src/hooks/useStore.ts

> import { useStore } from '@/hooks/useStore';

Reviewer: `@/hooks` or `@hook` or `@hooks`?
